### PR TITLE
Genotype concordance analysis, again

### PR DESCRIPTION
**Pull request by mariogiov**
Now with bonus fields and values that are accurate, genotype concordance output file parsing, and automatic updating of Charon when the genotype jobs finish. I need to add a script cron can use to check for incoming genotyping files.

### CODE REVIEW

File: ngi_pipeline/engines/piper_ngi/local_process_tracking.py [hunk 1]
```diff
@@ -81,7 +81,12 @@ def update_charon_with_local_jobs_status(quiet=False, config=None, config_file_p
             try:
                 if piper_exit_code == 0:
                     # 0 -> Job finished successfully
-                    set_status = "ANALYZED" # For the sample level
+                    if workflow == "merge_process_variantcall":
+                        status_field = "analysis_status"
+                        set_status = "ANALYZED" # sample level
```
Comment (mariogiov) [inline ngi_pipeline/engines/piper_ngi/local_process_tracking.py]: These differences between the various sample/seqrun and genotype_status/analysis_status field values are annoying and probably not necessary. We could consider making them more uniform? At least for values that mean the same thing -- `ANALYZED` means the same as `DONE`, for instance.

File: ngi_pipeline/engines/piper_ngi/local_process_tracking.py [hunk 2]
```diff
@@ -198,8 +197,10 @@ def update_charon_with_local_jobs_status(quiet=False, config=None, config_file_p
                         set_status = "UNDER_ANALYSIS"
                         if workflow == "merge_process_variantcall":
                             status_field = "alignment_status"
+                            recurse_status = "RUNNING"
```
Comment (mariogiov) [inline ngi_pipeline/engines/piper_ngi/local_process_tracking.py]: Here again is an annoying inconsistency. For `analysis_status`, it's `RUNNING` for seqrun records but `UNDER_ANALYSIS` for sample records. It is `UNDER_ANALYSIS` for both samples and seqruns with regard to the `genotype_status` field, which is nice.

File: ngi_pipeline/engines/piper_ngi/results_parsers.py [hunk 3]
```diff
@@ -33,3 +41,116 @@ def parse_qualimap_coverage(genome_results_file):
...
+def parse_genotype_concordance(genotype_concordance_file):
+    concordance_data = []
+    gt_values_list = []
+    with open(genotype_concordance_file, 'r') as f:
+        for line in iter(f.readline, ''):
+            header_location = None
+            if line.startswith("#:GATKTable:GenotypeConcordance_Summary"):
+                header_values = [h.strip() for h in f.readline().strip().split('  ') if h.strip()]
+                header_values = [h.lower().replace("-", "_").replace(" ", "_") for h in header_values]
+                GTValue = namedtuple('GTValue', header_values)
```
Comment (mariogiov) [inline ngi_pipeline/engines/piper_ngi/results_parsers.py]: Wasn't @b97pla doing something like this the other day? Something with a config file. `namedtupled` is the way. Maybe I already said that.

Comment (mariogiov) [inline ngi_pipeline/engines/piper_ngi/results_parsers.py]: Although I don't recall the particulars. Maybe `namedtuple` is the way, maybe it's not. Who the eff am I. I don't know.

File: ngi_pipeline/utils/charon.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        else:
+        # Alright then we'll do this the hard way. I'm looking at you, Uppsala
+        # This is probably better implemented within Charon. I'm looking at you, Denis
+        # Also this will just grab the first project containing the sample --
+        # if it exists in more than one project I don't know what to tell you
+        # except that your sample naming scheme sucks.
+            if not all_projects:
+                all_projects = [project["projectid"] for project in \
+                                charon_session.projects_get_all()['projects']]
+            for project_id in all_projects:
```
Comment (mariogiov) [inline ngi_pipeline/utils/charon.py]: @Galithil do you see what I'm doing here? It's laughably slow. And the whole time I'm waiting for it to finish I just sit thinking how dumb it is to do it this way. Can you make this happen on the database end somehow?

File: ngi_pipeline/engines/piper_ngi/results_parsers.py [hunk 5]
```diff
@@ -33,3 +41,116 @@ def parse_qualimap_coverage(genome_results_file):
...
+        mean_autosomal_coverage += parse_qualimap_coverage(genome_result)
+    return mean_autosomal_coverage
+
+
+
+def parse_genotype_concordance(genotype_concordance_file):
+    concordance_data = []
+    gt_values_list = []
+    with open(genotype_concordance_file, 'r') as f:
+        for line in iter(f.readline, ''):
```
Comment (mariogiov) [inline ngi_pipeline/engines/piper_ngi/results_parsers.py]: You're Not As Good At Python As You Thought You Were
or at least I wasn't. TIL: http://stackoverflow.com/a/14145118/1256058

File: ngi_pipeline/utils/charon.py (hunk 4 above)
Comment (Galithil) [inline ngi_pipeline/utils/charon.py]: So you would like a way to query the database ? No problem, maybe we could develop a Simple Query Language ... oh wait ...

Comment (Galithil) [inline ngi_pipeline/utils/charon.py]: does that work for you ?
https://github.com/NationalGenomicsInfrastructure/charon/pull/86

Comment (mariogiov) [inline ngi_pipeline/utils/charon.py]: It's perfect! I'll adjust and re-PR later today

File: ngi_pipeline/utils/charon.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    for sample_name in sample_list:
+        # First see if we can just parse out the project id from the sample name
+        m = STHLM_SAMPLE_RE.match(sample_name)
+        if m:
+            project_id = m.groups()[0]
+            projects_dict[project_id].add(sample_name)
+        else:
+            # Otherwise check all the projects for matching samples
+            owner_projects_list = charon_session.sample_get_projects(sample_name).get('projects')
```
Comment (mariogiov) [inline ngi_pipeline/utils/charon.py]: @Galithil is this correct? The structure of the JSON string returned, I mean. I can test it when you deploy the Charon PR but I just took a guess here.

Comment (mariogiov) [inline ngi_pipeline/utils/charon.py]: I'm assuming that if it returns  list, also.

File: ngi_pipeline/database/classes.py [hunk 7]
```diff
@@ -117,6 +116,9 @@ def sample_get(self, projectid, sampleid):
     def sample_get_libpreps(self, projectid, sampleid):
         return self.get(self.construct_charon_url('libpreps', projectid, sampleid)).json()
 
+    def sample_get_projects(self, sampleid):
+        return self.get(self.construct_charon_url('projectidsfromsampleid', sampleid)).json()
```
Comment (mariogiov) [inline ngi_pipeline/database/classes.py]: @Galithil I think this is correct as well but let me know if I've got it wrong.

Comment (Galithil) [inline ngi_pipeline/database/classes.py]: it's deployed and returns a list 100% of the time

**Comment by mariogiov** [comment]
@vezzi @Galithil I think this can be merged now. It updates Charon values and all.

**Pull request merged**